airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py

```diff
@@ -41,18 +122,581 @@ def create_component(self, model_type: Type[BaseModel], component_definition: Co
+    # component fields, this needs to be remapped manually
+    model_args["options"] = model_args.pop("$options", {})
+
+    # Pydantic is unable to parse a custom component's fields that are subcomponents into models because their fields and types are not
+    # defined in the schema. The fields and types are defined within the Python class implementation. Pydantic can only parse down to
+    # the custom component and this code performs a second parse to convert the sub-fields first into models, then declarative components
+    for model_field, model_value in model_args.items():
+        # If a custom component field doesn't have a type set, we try to use the type hints to infer the type
+        if isinstance(model_value, dict) and "type" not in model_value and model_field in component_fields:
+            derived_type = derive_component_type_from_type_hints(component_fields.get(model_field))
+            if derived_type:
+                model_value["type"] = derived_type
+
+        if _is_component(model_value):
+            model_args[model_field] = _create_nested_component(model, model_field, model_value, config)
+        elif isinstance(model_value, list):
+            vals = []
+            for v in model_value:
+                if isinstance(v, dict) and "type" not in v and model_field in component_fields:
+                    derived_type = derive_component_type_from_type_hints(component_fields.get(model_field))
+                    if derived_type:
+                        v["type"] = derived_type
+                if _is_component(v):
+                    vals.append(_create_nested_component(model, model_field, v, config))
+            model_args[model_field] = vals
+
+    kwargs = {class_field: model_args[class_field] for class_field in component_fields.keys() if class_field in model_args}
+    return custom_component_class(**kwargs)
+
+
+def derive_component_type_from_type_hints(field_type: str) -> Optional[str]:
+    interface = field_type
+    while True:
+        origin = get_origin(interface)
+        if origin:
+            # Unnest types until we reach the raw type
+            # List[T] -> T
+            # Optional[List[T]] -> T
+            args = get_args(interface)
+            interface = args[0]
+        else:
+            break
+    if isinstance(interface, type) and not is_builtin_type(interface):
+        return interface.__name__
+    return None
+
+
+def is_builtin_type(cls) -> bool:
+    if not cls:
+        return False
+    return cls.__module__ == "builtins"
+
+
+def _create_nested_component(model, model_field: str, model_value: Any, config: Config) -> Any:
+    type_name = model_value.get("type", None)
+    # if not type_name:
+    #     raise ValueError(
+    #         f"Error while parsing custom component {model.class_name}. Subcomponent field '{model_field}' should have a 'type' specified"
+    #     )
```

Comment:
Will remove this block later, but wanted to leave this in as a discussion point. Ideally we would like to flag any custom components that do not have a type, however we can't differentiate between a general dictionary mapping vs. a subcomponent that is missing it's type. If anyone has a clever idea for this, otherwise we just aren't able to surface the error quite as nicely as we would like

Reply:
Hmm so if `not type_name`, we'll basically always hit the `ValueError` on line 283. Seems to me that'd also be incorrect in the case of a dictionary. Is that right?

Reply:
Not that I'm aware of, but maybe I'm misunderstanding. I thought that was a situation you were referring to when you said "we can't differentiate between a general dictionary mapping vs. a subcomponent that is missing it's type".

Reply:
Not that I'm aware of @girarda but seems that it's theoretically possible and that's what we're protecting against - can you confirm @brianjlai?

Reply:
I haven't seen any usage of it so far, but it is not an inconceivable use case. Right now in our schema, the only non-component raw dictionary usage we have is for request body/headers/parameters (request_options_provider) and refresh_request_body (oauth authenticator).

Reply:
@clnoll that's a good catch... actually if we don't have a `type` we'll end up with `None`. My thoughts are that if type name ends up being None, then just return the original dictionary back, otherwise we continue to do the parse object resolution